lib/pbench/server/api/resources/query_apis/search_api.py

```diff
@@ -0,0 +1,224 @@
+                Parameter("user", ParamType.USER, required=False),
+                Parameter("start", ParamType.DATE, required=True),
+                Parameter("end", ParamType.DATE, required=True),
+                Parameter("search_term", ParamType.STRING, required=False),
+                Parameter("fields", ParamType.LIST, required=False),
+            ),
+        )
+
+    def flatten_json(self, es_json: Dict[AnyStr, Any]) -> Dict[AnyStr, Any]:
+        """
+        Helper function for Postprocess for flattening the json. Any exception raised here will be
+        delegated as a postprocess error.
+        We convert a nested data layer object into a new object with only one layer of key/value pairs.
+        Objects such as parent -> child are flattened as parent.child like following:
+        parent -> array -> [obj1,objN] are flattened as
+                parent.array.0 = obj1, parent.array.N = objN
+        parent -> array -> [{key1: value1}, {key2: value2}] are flattened as
+                parent.array.key1 = value1, parent.array.key2 = value2
+        parent -> array -> [[obj1, obj2], [obj3]] are flattened as
+                'parent.array.0.0': obj1, 'parent.array.0.1': obj2, 'parent.array.1.0': obj3
+        json_in =
+            {
+                    "@timestamp": "2021-07-12T14:08:46.563303",
+                    "run": {
+                        "controller": "dhcp31-171.perf.lab.eng.bos.redhat.com",
+                        "name": "pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T14.08.46"
+                    },
+                    "host_tools_info": [
+                        {
+                            "hostname": [{"test1":"dhcp31-171"}, {"test2":"dhcp31-195"}]
+                        }
+                    ]
+                }
+        json_out =
+            {
+                '@timestamp': '2021-07-12T14:08:46.563303',
+                'run.controller': 'dhcp31-171.perf.lab.eng.bos.redhat.com',
+                'run.name': 'pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T14.08.46',
+                'host_tools_info.hostname.test1': 'dhcp31-171',
+                'host_tools_info.hostname.test2': 'dhcp31-195'
+            }
+        """
+        if type(es_json) is not dict:
+            raise PostprocessError(
+                HTTPStatus.INTERNAL_SERVER_ERROR,
+                "Elasticsearch _source value is not a json",
+            )
+
+        def recurse(
+            es_json: Any, prefix: AnyStr, out: Dict[AnyStr, Any]
+        ) -> Dict[AnyStr, Any]:
+            if type(es_json) is dict:
+                for key in es_json:
+                    while prefix[-2:-1] == ".":
+                        prefix = prefix[:-1]
+                    recurse(es_json[key], prefix + key + ".", out)
+            elif type(es_json) is list:
+                for i, element in enumerate(es_json):
+                    if type(element) is dict:
+                        recurse(element, prefix + ".", out)
```

Comment:
I tried to experiment with what you suggested here but it produced some weird output. What _I think_ is happening is that that if the type of the element is `dict` we have to add a dot to separate the child and parent keys. What you see on line 86 is for each element of the list and on line 80 we check for each element of the dict and they both need to add a dot as a separation parameter. 

However, I think this is getting little complicated now so I have tried to clean this function a little bit and I am trying to follow a simple flattening rule that you also mentioned earlier here https://github.com/distributed-system-analysis/pbench/pull/2381#discussion_r680244798 I think that would be good to follow for now.

So the following json
```
{
    "@timestamp": "2021-07-12T14:08:46.563303",
    "run": {
        "controller": "dhcp31-171.perf.lab.eng.bos.redhat.com",
        "name": "pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T14.08.46"
    },
    "host_tools_info": [
        {
            "hostname": [{"test":"dhcp31-171"}, {"test2":"dhcp31-195"}]
        }
    ]
}
```
would reformat like this 
```
{
'@timestamp': '2021-07-12T14:08:46.563303', 
'run.controller': 'dhcp31-171.perf.lab.eng.bos.redhat.com', 
'run.name': 'pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T14.08.46', 
'host_tools_info.0.hostname.0.test': 'dhcp31-171', 
'host_tools_info.0.hostname.1.test2': 'dhcp31-195'
}
```
If the value is list for any key in our dictionary then irrespective of the type of the element in the list we would add an index prefix like 0 or 1. This simplifies the code a little bit and also avoid any overlap (Although as we discussed this might be more confusing for the reader without fully understanding what this means).

Reply:
Okay discard the above comment @webbnh sorry about that. I did more experiment and playing around with what you suggested and it seems to be working fine, we dont need a dot here apparently and that helped to get rid of code on line 81-82.